x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -727,16 +814,332 @@ public void runAsDenied(String requestId, Authentication authentication, RestReq
+                    .endObject() // user
+                    .endObject() // password
+                    .endObject();
+            logEntry.with(CHANGE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(PutRoleRequest putRoleRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "put_role");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                        .startObject("role")
+                        .field("name", putRoleRequest.name())
+                        // the "privilege" nested structure, where the "name" is left out, is closer to the event structure
+                        // for creating API Keys
+                        .startObject("privilege")
+                            .startArray("index");
+                                for (RoleDescriptor.IndicesPrivileges indicesPrivileges : putRoleRequest.indices()) {
+                                    withIndicesPrivileges(builder, indicesPrivileges);
+                                }
+                            builder.endArray() // index
+                            .array("cluster", putRoleRequest.cluster())
+                            .array("run_as", putRoleRequest.runAs())
+                            // the toXContent method of the {@code RoleDescriptor.ApplicationResourcePrivileges) does a good job
+                            .field("application", putRoleRequest.applicationPrivileges())
+                            .field("global");
+                            // This fails if this list contains multiple instances of the {@code ManageApplicationPrivileges}
+                            // Again, only the transport client can produce this, and this only introduces a different failure mode and
+                            // not a new one (i.e. without auditing it would fail differently, but it would still fail)
+                            ConfigurableClusterPrivileges.toXContent(builder, ToXContent.EMPTY_PARAMS,
+                                    Arrays.asList(putRoleRequest.conditionalClusterPrivileges()));
+                            if (putRoleRequest.metadata() != null && false == putRoleRequest.metadata().isEmpty()) {
+                                // JSON building for the metadata might fail when encountering unknown class types.
+                                // This is NOT a problem because such metadata (eg containing GeoPoint) will most probably
+                                // cause troubles in downstream code (eg metadata store), so this simply introduces a new failure mode.
+                                // Also the malevolent metadata can only be produced by the transport client.
+                                builder.field("metadata", putRoleRequest.metadata());
+                            }
+                        builder.endObject() // privilege
```

Comment:
As is the case with auditing all the other requests in this PR, I have chosen the field names under which we audit for **consistency in the audit context**. You might recall that we had the option to audit low level indexing requests, or high level transport requests. But we've chosen against those because there were cases where those approaches would expose internal fluff.

All this long PR is about is **mapping fields in request bodies to fields in audit events**. That is all, no more, no less.

In this particular case, the `PutUserRequest` and `CreateApiKeyRequest` both specify some form of role descriptor that must be updated or created. I think the goal of this PR is to make them look similarly when audited.
In addition, I don't think we should audit transient metadata. Still, when a request field is null or empty we have the option to skip the field in the audit, and I've done so for example in the FLS/DLS and metadata cases where it would've created noise, IMO.

This is **my** rationale behind these naming decisions.
**If you have any suggestion, I will gleefully accept it**, as I believe there is some point beyond which the utility of a better looking log is not worth the debate it entails.

Reply:
> I don't think we should audit transient metadata. Still, when a request field is null or empty we have the option to skip the field in the audit
>
I get it that we strive to be more concise when serialising the `RoleDescriptor` for auditing. We can do these while keeping the existing field names. I am not suggesting to just reuse the `RoleDescriptor#toXContent` method, it's just about keeping the same field name when the field needs to be serialised in this place.

> I have chosen the field names under which we audit for consistency in the audit context. You might recall that we had the option to audit low level indexing requests, or high level transport requests. But we've chosen against those because there were cases where those approaches would expose internal fluff.
>
RoleDescriptor is not really that internally. In many cases, users need to specify them in the request body. So I'd expect some familarity from users and using the same field names has the benefit of maintaining that familarity. 
It's good to have consistency among different requests for auditing, maybe even consistent with ECS to some extend. I don't think it's creating any inconsistency issue by reusing the same field names from `RoleDescriptor`.

Reply:
I agree with @ywangd here.

The get role API returns a structure like:
```
{
  "role_name": {
    "cluster": [ "monitor" ],
    "indices": [
      {
        "names": [ "abc" ],
        "privileges": [ "read" ],
        "field_security": {
          "grant": [  "*" ],
          "except": [ "private.*" ]
        },
        "query": "{ \"term\": { \"public\": true}}",
        "allow_restricted_indices": false
      }
    ],
    "applications": [],
    "run_as": [],
    "metadata": {},
    "transient_metadata": {
      "enabled": true
    }
  }
}
``` 

But, I understand the code correctly, we will audit this as:
```
"role": {
  "name": "role_name",
  "privilege": {
    "cluster": [ "monitor" ],
    "indices": [
      {
        "names": [ "abc" ],
        "privileges": [ "read" ],
        "fields": {
          "grant": [  "*" ],
          "except": [ "private.*" ]
        },
        "documents":{ "query_string": "{ \"term\": { \"public\": true}}" }
      }
    ],
    "applications": [],
    "run_as": [],
    "metadata": {}
  }
}
```

Not auditing `transient_metadata` makes sense, but what's the justification for using different field names for `"field_security"` vs `"fields"` and `"query"` vs `"documents": { "query_string":`

> I think the goal of this PR is to make them look similarly when audited.

Surely the more important goal is to make them _understandable_ to people reading the audit logs, and using the same terms that we use in the API Keys & Roles APIs does a better job of that.
Where API keys and Roles have differences, then we can talk about how to make them consistent in the audit logs, but having the audit logs be arbitrarily different from the existing request/response bodies for both Roles and API keys doesn't seem helpful.

Reply:
In terms of specific suggestions, I propose:
- If we want to have a nested object for the role descriptor (what is under the `"privilege"` field in the example above) then we should follow the naming from the [Create API Key](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/security-api-create-api-key.html) request, and use some variation on `role_descriptors` (which can be `role_descriptor` if it's not going to be an array).
- Use `indices`, `field_security`, `applications`
- Also use `query`, without any nested object, and output it as a String, as you already do.
- I'm ambivalent about whether to audit `"allow_restricted_indices": false`. The API always shows it (true or false), but that doesn't mean audit needs to
- Probably skip `global` if it's empty, as the API does, and continue to skip empty `metadata` as you do above (but again, I'm mostly ambivalent)
- Order the fields in the same order that RoleDescriptor does (because consistency is better)

Reply:
I have pushed f27629e02bf9264769464c154ee94a to address all the above suggestions.

Two minor mentions related to the suggestions above:
* when putting a role or creating an API key, `index` and `application` (singulars) can also be used (as in the create api key example https://www.elastic.co/guide/en/elasticsearch/reference/7.10/security-api-create-api-key.html), but in responses they are always plural. Given your suggestion they are now audited with plurals only as well.
* I maintain that `query_string` is a better name to `query` for the DLS query string. I _personally_ think replicating the structure of the API request or response is the wrong trade-off in this specific case.

One important note, which is more slightly related to the suggestions above:
The audited role descriptors don't show their names in the create api key case. Those names are not important and are not audited in the other cases too, and it helps with the consistency with the other audited events (having a single `name` field). See for example the `role_descriptor: {` and `role_descriptors: [` in the below samples:  
```
{
  "type": "audit",
  "timestamp": "2020-12-09T15:29:32,693+0200",
  "node.id": "vnYUc3fmSm2QyULf6t9-Sg",
  "event.type": "security_config_change",
  "event.action": "put_role",
  "request.id": "TsYX5M8WQR6hcOC_7yBskQ",
  "put": {
    "role": {
      "name": "role_name",
      "role_descriptor": {
        "cluster": [
          "monitor"
        ],
        "indices": [
          {
            "names": [
              "abc"
            ],
            "privileges": [
              "read"
            ],
            "field_security": {
              "except": [
                "private.*"
              ]
            },
            "query": "{\"match\": {\"title\": \"foo\"}}"
          }
        ],
        "applications": [
          {
            "application": "myapp",
            "privileges": [
              "admin",
              "read"
            ],
            "resources": [
              "*"
            ]
          }
        ],
        "run_as": [
          "other_user"
        ]
      }
    }
  }
}
```
```
{
  "type": "audit",
  "timestamp": "2020-12-09T15:29:57,264+0200",
  "node.id": "vnYUc3fmSm2QyULf6t9-Sg",
  "event.type": "security_config_change",
  "event.action": "create_apikey",
  "request.id": "_dwdVsKpRjaCZMyDS86XpQ",
  "create": {
    "apikey": {
      "name": "api_key_name",
      "expiration": "1d",
      "role_descriptors": [
        {
          "cluster": [
            "monitor"
          ],
          "indices": [
            {
              "names": [
                "abc"
              ],
              "privileges": [
                "read"
              ],
              "field_security": {
                "except": [
                  "private.*"
                ]
              },
              "query": "{\"match\": {\"title\": \"foo\"}}"
            }
          ],
          "applications": [
            {
              "application": "myapp",
              "privileges": [
                "admin",
                "read"
              ],
              "resources": [
                "*"
              ]
            }
          ],
          "run_as": [
            "other_user"
          ]
        }
      ]
    }
  }
}
```
In general, objects in the audit events **don't have variables as keys**.
I'm pointing this out because it is different from the responses of the GET APIs that can return multiple values.
But our "set" APIs, which are the ones that this PR is concerned with, only act on single objects, and
I think that having a `"name": "dummy", ...` field is more straightforward and **easier to parse** than keying with the value of the `name` (or equivalent) attribute, eg `"enabled": {"user": {"name": "test"}}` vs `"enabled": {"user": "test"}`.
Let me know if we need to discuss this further.

Reply:
The new structure looks fine to me. 

> when putting a role or creating an API key, index and application (singulars) can also be used
>
This is probably a legacy leniency issue and not something to be replicated. +1 to always use the plural format.

> The audited role descriptors don't show their names in the create api key case
> 
I think this is a good change. The name is not an inherent part of RoleDescriptor. For API keys, they are even less relevant. We don't report them in authenticate call anymore. So it makes sense to skip them. In fact, I like how the `role_descriptors` is an array instead of map in the audit. Honestly I think this should have been the format used by the `PutApiKeyRequest`. But that's another story.

> In general, objects in the audit events don't have variables as keys.
> I'm pointing this out because it is different from the responses of the GET APIs that can return multiple values.
But our "set" APIs, which are the ones that this PR is concerned with, only act on single objects, and
I think that having a "name": "dummy", ... field is more straightforward and easier to parse than keying with the value of the name (or equivalent) attribute, eg "enabled": {"user": {"name": "test"}} vs "enabled": {"user": "test"}.
>
I think I understand you point here, but the last example does not seem to align with what you described? No keys are variables in both format?